ARROW-3128: [C++] Support system shared zlib

kou opened the pull request:
Debian package recommends not static linking zlib for security reason. If we use zlib as a static library, Debian package lint reports the following error:

    E: libarrow10: embedded-library usr/lib/x86_64-linux-gnu/libarrow.so.10.0.0: zlib

embedded-library error detail: https://lintian.debian.org/tags/embedded-library.html

zlib detection by pkg-config is also added because system zlib provides zlib.pc.

Detect order (default):

  * pkg-config (as shared library)
  * System default path (as shared library)
  * Vendoered zlib (as static library)

Detect order with ZLIB_HOME environment variable:

  * zlib at ZLIB_HOME environment variable value (as shared library)

This is not backward compatible. We used vendored zlib by default.

Summary:

  * Use system zlib by default
    * Changed.
  * System zlib is searched by `pkg-config` or from the default system path
  * Use the specified zlib when `ZLIB_HOME` is specified
    * Not changed.
  * If `ZLIB_HOME` is specified, system zlib isn't searched by `pkg-config`
    * Not changed.
  * Use vendored zlib as fallback (no `ZLIB_HOME` and no system zlib)
    * Changed. Vendored zlib was the default.
  * Use system zlib as shared library
    * Changed. Zlib was always used as static library.
  * Use `ZLIB_HOME` zlib as shared library
    * Changed. Zlib was always used as static library.
  * Use vendored zlib as static library
    * Not changed.
  * Bundle zlib into wheels for Windows.
    * Changed.
    * Zlib isn't bundled into wheels for manylinux1 and macosx because we can assume zlib exists on these platforms.
    * See also:
      * https://mail.python.org/mm3/archives/list/distutils-sig@python.org/thread/ZZG6GL3XTBLBJXSITYHEXMFKN43EREB7/
      * https://mail.python.org/mm3/archives/list/distutils-sig@python.org/message/HLX65TU6U7N3JTUU4YRLBQNSFKZTG52X/
  * Build directory is always removed on AppVeyor.
    * Changed.
    * Because CMake's cache reports wrong result.
  * Use zlib in the build target Python directory instead of `ARROW_BUILD_TOOLCHAIN` on Travis CI.
    * Changed.
    * Because linking with zlib in `ARROW_BUILD_TOOLCHAIN` causes wrong `libpython.dylib` load on macOS. See comment in `ci/travis_script_python.sh` for details.
  * `after_failure` change in `.travis.yml` is a by-product of debugging crash on macOS.
    * Should I separate this change to another pull request?
  * `ZLIB_LIBS` has been removed.
    * Because it's not a good name and we don't use anywhere.
    * It doesn't contain library paths. It contains directories that includes library.

kou commented:
I'll work on changing .deb/.rpm to use this features after this change is merged.

pitrou commented on cpp/cmake_modules/FindZLIB.cmake:
```diff
@@ -39,45 +39,76 @@ elseif ( ZLIB_HOME )
-    find_path( ZLIB_INCLUDE_DIR NAMES zlib.h )
-    # Only look for the static library
-    find_library( ZLIB_LIBRARIES NAMES libz.a zlib )
+  pkg_check_modules(ZLIB zlib)
```
This uses pkg-config?

pitrou replied:
To make things clearer, perhaps we could use a different prefix for the pkg-config results, for example PKG_ZLIB. That would make the below more readable, IMHO.

kou replied:
It makes sense. I'll do.

pitrou commented on cpp/cmake_modules/FindZLIB.cmake:
```diff
@@ -39,45 +39,76 @@ elseif ( ZLIB_HOME )
+  set(ZLIB_LIBS)
+  if (ZLIB_STATIC_LIB)
+    get_filename_component(ZLIB_STATIC_LIB_DIR ${ZLIB_STATIC_LIB} DIRECTORY)
+    list(APPEND ZLIB_LIBS ${ZLIB_STATIC_LIB_DIR})
```
So ZLIB_LIBS lists directories? Uh :-/

kou replied:
We define the following API in the header of this file...

> #  ZLIB_LIBS, directory containing zlib libraries

I think that nobody uses `ZLIB_LIBS`. Can we remove `ZLIB_LIBS`?

pitrou commented on cpp/cmake_modules/FindZLIB.cmake:
```diff
@@ -39,45 +39,76 @@ elseif ( ZLIB_HOME )
+
 if (ZLIB_FOUND)
   if (NOT ZLIB_FIND_QUIETLY)
     if (PARQUET_MINIMAL_DEPENDENCY)
```
PARQUET_MINIMAL_DEPENDENCY isn't used anymore above, it seems, so perhaps we needn't test it here.

kou replied:
It may be used when we merge perquet-cpp into Arrow repository.
But "found the ZLIB shared library" message will be enough. We'll not need "Found the ZLIB header" message.
I'll remove this condition.

@xhochy Do you have any concerns?

xhochy replied:
👍

kou replied:
OK!

pitrou commented on cpp/cmake_modules/FindZLIB.cmake:
```diff
@@ -39,45 +39,76 @@ elseif ( ZLIB_HOME )
+	message(STATUS "Found the ZLIB shared library: ${ZLIB_SHARED_LIB}")
+      endif ()
     endif ()
   endif ()
```
Should we add a warning message if the zlib isn't found? Especially as we vendor it...

kou replied:
It should be done in `ThirdpartyToolchain.cmake` not in `FindZLIB.cmake`.
We don't care whether vendored or not here.

pitrou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -671,6 +671,18 @@ if (ARROW_WITH_ZLIB)
 # ZLIB
 
   if("${ZLIB_HOME}" STREQUAL "")
+    if(ARROW_ZLIB_VENDORED)
```
IMHO, we should make it an error to set both ZLIB_HOME and ARROW_ZLIB_VENDORED. It's already annoying enough to debug configuration issues.

kou replied:
I've removed `ARROW_ZLIB_VENDORED` to use system zlib by default and vendored zlib as fallback.

pitrou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -671,6 +671,18 @@ if (ARROW_WITH_ZLIB)
+  if(ZLIB_FOUND)
+    set(ZLIB_VENDORED 0)
+  else()
+    set(ZLIB_VENDORED 1)
```
Similarly, we should make it an error if ARROW_ZLIB_VENDORED is false and a third-party zlib couldn't be found.

pitrou replied:
Actually, this sounds reasonable since building needs the development headers, which aren't always installed (as opposed to the library itself, which is pretty much everywhere).

kou commented on cpp/CMakeLists.txt:
```diff
@@ -235,6 +236,14 @@ Pass multiple labels by dividing with semicolons")
     ON)
 
+  option(ARROW_ZLIB_VENDORED
+    "Use vendored Boost instead of existing Boost"
```
Oh... Sorry for my low quality pull request...
I'll fix it.

kou commented on cpp/cmake_modules/FindZLIB.cmake:
```diff
@@ -39,45 +39,76 @@ elseif ( ZLIB_HOME )
 
+set(ZLIB_LIB_NAME z)
+if (MSVC)
+  if (NOT ZLIB_MSVC_STATIC_LIB_SUFFIX)
```
OK.
I'll add a comment that zlib uses `zlib.lib` on Windows.
FYI: https://github.com/madler/zlib/blob/master/win32/README-WIN32.txt#L50

kou commented:
Now, CI is green again. Can someone review this?

Summary:

  * Use system zlib by default
    * Changed.
  * System zlib is searched by `pkg-config` or from the default system path
  * Use the specified zlib when `ZLIB_HOME` is specified
    * Not changed.
  * If `ZLIB_HOME` is specified, system zlib isn't searched by `pkg-config`
  * Use vendored zlib as fallback (no `ZLIB_HOME` and no system zlib)
    * Changed. Vendored zlib was the default.
  * Use system zlib as shared library
    * Changed. Zlib is always used as static library.
  * Use `ZLIB_HOME` zlib as static library
    * Not changed.
    * I gave up to use `ZLIB_HOME` zlib as shared library... There are two problems:
      * pyarrow wheel need to depend on zlib package:
        https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/build/1.0.7630/job/sumy5r26q6ex7va1#L6772
      * `python-test` is crashed only on macOS:
        Backtrace: https://travis-ci.org/apache/arrow/jobs/422620234#L2677-L2696
        * I don't know why...
  * Use vendored zlib as static library
    * Not changed.
  * `after_failure` change in `.travis.yml` is a by-product of debugging crash on macOS.
    * Should I separate this change to another pull request?
  * `ZLIB_LIBS` has been removed.
    * Because it's not a good name and we don't use anywhere.
    * It doesn't contain library paths. It contains directories that includes library.

wesm commented on cpp/CMakeLists.txt:
```diff
@@ -603,7 +604,11 @@ if (ARROW_WITH_SNAPPY)
 
 if (ARROW_WITH_ZLIB)
-  SET(ARROW_STATIC_LINK_LIBS zlib ${ARROW_STATIC_LINK_LIBS})
+  if (ZLIB_SHARED_LIB)
```
We still need a flag to link statically since we deploy Python and C++ packages on platforms which may not have libz in the dynamic loader path

xhochy replied:
Is this static linking required on Windows? OSX and Linux don't need it in wheels.

wesm replied:
I just looked at a default install of Python 3.6 on Windows 64 and it doesn't appear to include zlib. I tried doing `ctypes.CDLL('z.dll')` and `CDLL('libz.dll')` and neither work. So either we need to bundle the shared library in wheels or statically link

kou replied:
OK.
I'll try to bundle zlib as shared library for wheels.
If it's succeeded, I'll remove static link support for `ZLIB_ROOT` case.

wesm commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -671,6 +671,17 @@ if (ARROW_WITH_ZLIB)
+      ADD_THIRDPARTY_LIB(zlib SHARED_LIB ${ZLIB_SHARED_LIB})
+    else()
+      ADD_THIRDPARTY_LIB(zlib STATIC_LIB ${ZLIB_STATIC_LIB})
+    endif()
```
This will need changing per comment above

wesm commented on cpp/src/arrow/util/CMakeLists.txt:
```diff
@@ -54,7 +54,11 @@ endif()
+  ADD_ARROW_TEST(compression-test EXTRA_LINK_LIBS ${ZLIB_SHARED_LIB})
+else()
+  ADD_ARROW_TEST(compression-test EXTRA_DEPENDENCIES zlib)
+endif()
```
Shouldn't this be exposed as a transitive dependency of arrow_static/arrow_shared? I think there's some way to get CMake to automatically include transitive link dependencies

xhochy commented:
zlib is not part of the `manylinux1` spec, so at least there, we should ship it with Arrow.

pitrou commented:
> zlib is not part of the manylinux1 spec, so at least there, we should ship it with Arrow.

I had missed that. It's a bit surprising, I wonder if there are GNU/Linux systems without even zlib installed.

pitrou commented:
I've asked a question about zlib here: https://mail.python.org/mm3/archives/list/distutils-sig@python.org/thread/ZZG6GL3XTBLBJXSITYHEXMFKN43EREB7/

kou commented:
CI is green again. Can someone review this again?

Summary:

  * Use system zlib by default
    * Changed.
  * System zlib is searched by `pkg-config` or from the default system path
  * Use the specified zlib when `ZLIB_HOME` is specified
    * Not changed.
  * If `ZLIB_HOME` is specified, system zlib isn't searched by `pkg-config`
    * Not changed.
  * Use vendored zlib as fallback (no `ZLIB_HOME` and no system zlib)
    * Changed. Vendored zlib was the default.
  * Use system zlib as shared library
    * Changed. Zlib was always used as static library.
  * Use `ZLIB_HOME` zlib as shared library
    * Changed. Zlib was always used as static library.
  * Use vendored zlib as static library
    * Not changed.
  * Bundle zlib into wheels for Windows.
    * Changed.
    * Zlib isn't bundled into wheels for manylinux1 and macosx because we can assume zlib exists on these platforms.
    * See also:
      * https://mail.python.org/mm3/archives/list/distutils-sig@python.org/thread/ZZG6GL3XTBLBJXSITYHEXMFKN43EREB7/
      * https://mail.python.org/mm3/archives/list/distutils-sig@python.org/message/HLX65TU6U7N3JTUU4YRLBQNSFKZTG52X/
  * Build directory is always removed on AppVeyor.
    * Changed.
    * Because CMake's cache reports wrong result.
  * Use zlib in the build target Python directory instead of `ARROW_BUILD_TOOLCHAIN` on Travis CI.
    * Changed.
    * Because linking with zlib in `ARROW_BUILD_TOOLCHAIN` causes wrong `libpython.dylib` load on macOS. See comment in `ci/travis_script_python.sh` for details.
  * `after_failure` change in `.travis.yml` is a by-product of debugging crash on macOS.
    * Should I separate this change to another pull request?
  * `ZLIB_LIBS` has been removed.
    * Because it's not a good name and we don't use anywhere.
    * It doesn't contain library paths. It contains directories that includes library.

wesm commented:
Thanks @kou for working on this! I will review today or tomorrow. I added your summary to the PR description so it will end up in the commit message

kou commented on ci/travis_script_python.sh:
```diff
@@ -35,6 +35,15 @@ CONDA_ENV_DIR=$TRAVIS_BUILD_DIR/pyarrow-test-$PYTHON_VERSION
 conda create -y -q -p $CONDA_ENV_DIR python=$PYTHON_VERSION cmake curl
 conda activate $CONDA_ENV_DIR
 
+# We should zlib in the target Python directory to avoid loading wrong
```
Oh, I've fixed this.

wesm commented on cpp/src/arrow/util/CMakeLists.txt:
```diff
@@ -54,7 +54,7 @@ endif()
 ADD_ARROW_TEST(bit-util-test)
 ADD_ARROW_TEST(checked-cast-test)
-ADD_ARROW_TEST(compression-test)
+ADD_ARROW_TEST(compression-test EXTRA_LINK_LIBS zlib)
```
Seems like this should transitively linked as a result of linking to libarrow.a. We should make a follow up to make sure that CMake is set to include transitive dependencies when statically linking

kou replied:
If we add `zlib` to `ARROW_STATIC_LINK_LIBS` even when we use zlib as shared library, we can remove this: https://github.com/apache/arrow/pull/2483/commits/db17bdb4c00ee38f0ea09fe7e4fc881290a13e5a

I'm not sure that it's a good manner in this project...

wesm replied:
OK

wesm commented on python/CMakeLists.txt:
```diff
@@ -263,6 +263,66 @@ function(bundle_boost_lib library_path)
 endfunction()
 
+function(bundle_zlib zlib_home)
+  if (MSVC)
```
Can FindZLIB.cmake not be used here?

kou replied:
We can't use FindZLIB.cmake here.
It finds `zlib.lib` to link zlib as shared library. But we need `zlib.dll` here.

Anyway, I've removed unused codes by https://github.com/apache/arrow/pull/2483/commits/3db0830345d865160d493d10308310b41ee2aa44
The removed codes are for non Windows. I kept them because we may use them later. But we can remove unused codes.

wesm replied:
OK, no problem. Thank you

kou commented:
Thanks.
#2554 will solve this failure.
